spmd/tensor/dispatch.py

```diff
@@ -0,0 +1,73 @@
+from dataclasses import dataclass
+from typing import Any, Callable, Dict
+
+from spmd.tensor.utils import (
+    wrap,
+)
+
+"""
+If set to true, __DEBUG_STRICT will fail when an op doesn't have a sharding rule registered.
+"""
+_DEBUG_STRICT = False
+
+
+@dataclass
+class OpInfo(object):
+    op_call: Callable
+    args_with_local_tensor: Any
+    kwargs_with_local_tensor: Any
+    args_spec: Any
+    kwargs_spec: Any
+
+
+def dispatch_operator(
+    op_info: OpInfo,
+    op_to_rules: Dict[str, Callable],
+) -> Any:
+
+    op_key = str(op_info.op_call)
+    sharding_prop_func = op_to_rules.get(op_key, None)
+    if sharding_prop_func is not None:
+        # step 1. there's sharding propagation rule, run
+        # sharding propagation to get output placements
+        output_placements = sharding_prop_func(op_info)
+
+        # step 2. if can't get output placement (i.e.
+        # no ruls apply for input placements), we do
+        # auto redistribute on inputs to get an
+        # eligble input, which we will pick a target
+        # placement base on the redistribute cost
+        # TODO: implement auto distribute with simple
+        # cost estimation
+        if output_placements is None:
+            # do auto distributed/boxing here
+            raise NotImplementedError("auto redistribute not implemented yet!")
+
+        # run local op computation
+        local_results = op_info.op_call(
+            *op_info.args_with_local_tensor, **op_info.kwargs_with_local_tensor
```

Comment:
@wanchaol I remember in our discussion we had mentioned that kwargs may need to be updated by the sharding propagation, is this something that we expect to happen in-place inside of the rules? Maybe we should clarify that , or even return a new set of kwargs from the prop. rule to make it explicit?

Reply:
@wanchaol Another thing -- maybe a suggestion -- there's a tradeoff of generality vs. how much you can do on the sharding propagation function depending on how much info you pass into it. 

Specifically, in order for us to be able to support ahead of time propagation, would it be a good thing if we made surre to only pass "meta" tensors or "fake" tensors into the sharidn gpropagation function , so to make sure its implementaiton would never rely on .data?

Reply:
chatted offline: We are actually only updating the “non-tensor” args or “non-tensor” kwargs, which means we don’t even need the “tensor” args/kwargs metadata (i.e. `tensor.view(*shape)`, no need for tensor metadata like size, we only need to update the *shape tuple to update the local tensor shape). The only reason we keep “tensor” args in `args_with_local_tensor`, is that after the modification on “non-tensor” args, we could directly pass `args_with_local_tensor` to the local op computation `op_info.op_call(*op_info.args_with_local_tensor, **op_info.kwargs_with_local_tensor)`. We will need to figure out a way to disallow modification of "tensors" data on `args_with_local_tensor` with each sharding propagation rules.

Reply:
Makes sense. Just note that we should if possible disallow _read_ access to .data as well, otherwise we won't be able to do ahead-of-time sharding propagation.

Reply:
I just updated the PR with OpSchema, where I explicitly document "what can be used" and "what can be changed" as propagation rule author. "tensor" argument should be explicitly disallowed, but we didn't have a way to check if it's getting modified or not. We can take this as a follow up and make sure there's no way to modify/read tensor data in propagation rules https://github.com/pytorch/PiPPy/issues/336

regarding on disallow read access of .data, could you elaborate a bit more on it, is it because ahead of time sharding propagation likely won't have any .data materialized?